File: skills/cumcm-code-writer/references/algorithm_library/02_评价决策/rsr/matlab/main.m
% 秩和比 RSR · 综合评价与分档（MATLAB 最小可运行模板）
here = fileparts(mfilename('fullpath'));
addpath(fullfile(here, '..', '..', '..', '_common', 'matlab'));

X = [92 0.83 3.1 8; 78 0.71 4.5 6; 85 0.90 2.7 9; 70 0.64 5.2 5; 88 0.79 3.6 7];
positive = [true true false true];
w = [0.3 0.3 0.2 0.2];
n = size(X, 1); m = size(X, 2);

R = zeros(n, m);
for j = 1:m
    col = X(:, j);
    if ~positive(j)
        col = -col;
    end
    [~, idx] = sort(col);
    r = zeros(n, 1);
    r(idx) = 1:n;
    R(:, j) = r;
end
rsr = (R * w') / n;
[~, order] = sort(rsr, 'descend');
rank = zeros(n, 1); rank(order) = 1:n;

mmlib_report('秩和比', sprintf('RSR=%s', mat2str(round(rsr', 4))), sprintf('排序=%s', mat2str(rank')));
T = table("A" + string(1:n)', rsr, rank, 'VariableNames', {'方案', 'RSR', '排序'});
mmlib_savetable(here, 'rsr', T);

File: skills/cumcm-code-writer/references/algorithm_library/02_评价决策/rsr/python/main.py
"""秩和比 RSR · 综合评价与分档（Python 最小可运行模板）

步骤：指标同向化 → 编秩（高优指标从小到大编秩）→ 加权秩和比 RSR → 按分位数分档。
接口：solve(X, positive, weight) -> (RSR, 排序, 档位)；运行：python main.py
"""
from __future__ import annotations

import sys
from pathlib import Path

import numpy as np
import pandas as pd

sys.path.insert(0, str(Path(__file__).resolve().parents[3] / "_common" / "python"))
from mmlib import demo, report, save_table  # noqa: E402


def build_demo() -> dict:
    X = np.array([[92, 0.83, 3.1, 8], [78, 0.71, 4.5, 6], [85, 0.90, 2.7, 9],
                  [70, 0.64, 5.2, 5], [88, 0.79, 3.6, 7]], dtype=float)
    return {"X": X, "positive": [True, True, False, True], "weight": [0.3, 0.3, 0.2, 0.2]}


def solve(X, positive, weight):
    X = np.asarray(X, float)
    weight = np.asarray(weight, float)
    n, m = X.shape
    ranks = np.zeros((n, m))
    for j in range(m):
        col = X[:, j] if positive[j] else -X[:, j]
        ranks[:, j] = pd.Series(col).rank(method="average").to_numpy()
    rsr = (ranks * weight).sum(axis=1) / n
    order = (-rsr).argsort().argsort() + 1
    # 按 RSR 排序后按 30%/40%/30% 数量比例分档
    sorted_index = np.argsort(rsr)
    grades = np.empty(n, dtype=object)
    cuts = np.cumsum([0.3, 0.4, 0.3]) * n
    labels = ["三档（差）", "二档（中）", "一档（优）"]
    for index, position in enumerate(sorted_index):
        grades[position] = labels[int(np.searchsorted(cuts, index + 0.5, side="right"))]
    return rsr, order, grades


if __name__ == "__main__":
    data = demo(__file__, build_demo)
    rsr, order, grades = solve(data["X"], data["positive"], data["weight"])
    report("秩和比", RSR=np.round(rsr, 4).tolist(), 排序=order.tolist())
    save_table(__file__, "RSR评价", {"方案": [f"A{i + 1}" for i in range(len(rsr))],
                                     "RSR": np.round(rsr, 6), "排序": order, "档位": grades})
